test/cpp/ext/otel/otel_tracing_test.cc

```diff
@@ -0,0 +1,545 @@
+    ASSERT_TRUE(OpenTelemetryPluginBuilder()
+                    .SetTracerProvider(std::move(tracer_provider))
+                    .SetTextMapPropagator(
+                        grpc::experimental::MakeGrpcTraceBinTextMapPropagator())
+                    .BuildAndRegisterGlobal()
+                    .ok());
+    grpc::ServerBuilder builder;
+    int port;
+    // Use IPv4 here because it's less flaky than IPv6 ("[::]:0") on Travis.
+    builder.AddListeningPort("0.0.0.0:0", grpc::InsecureServerCredentials(),
+                             &port);
+    builder.RegisterService(&service_);
+    server_ = builder.BuildAndStart();
+    server_address_ = absl::StrCat("localhost:", port);
+    auto channel = grpc::CreateChannel(server_address_,
+                                       grpc::InsecureChannelCredentials());
+    stub_ = EchoTestService::NewStub(channel);
+  }
+
+  void TearDown() override {
+    server_->Shutdown();
+    grpc_shutdown_blocking();
+    grpc_core::ServerCallTracerFactory::TestOnlyReset();
+    grpc_core::GlobalStatsPluginRegistryTestPeer::
+        ResetGlobalStatsPluginRegistry();
+  }
+
+  void SendRPC() {
+    EchoRequest request;
+    request.set_message("foo");
+    EchoResponse response;
+    grpc::ClientContext context;
+    grpc::Status status = stub_->Echo(&context, request, &response);
+  }
+
+  opentelemetry::nostd::shared_ptr<opentelemetry::trace::Tracer> tracer_;
+  std::shared_ptr<opentelemetry::exporter::memory::InMemorySpanData> data_;
+  CallbackTestServiceImpl service_;
+  std::string server_address_;
+  std::unique_ptr<grpc::Server> server_;
+  std::unique_ptr<EchoTestService::Stub> stub_;
+};
+
+TEST_F(OTelTracingTest, Basic) {
+  SendRPC();
+  absl::SleepFor(absl::Milliseconds(500 * grpc_test_slowdown_factor()));
+  auto spans = data_->GetSpans();
+  SpanData* client_span;
+  SpanData* attempt_span;
+  SpanData* server_span;
+  // Verify that we get 3 spans -
+  // 1) Client RPC Span - Sent.grpc.testing.EchoTestService/Echo
+  // 2) Attempt Span - Attempt.grpc.testing.EchoTestService/Echo
+  // 3) Server RPC Span - Recv.grpc.testing.EchoTestService/Echo
+  EXPECT_EQ(spans.size(), 3);
+  for (const auto& span : spans) {
+    EXPECT_TRUE(span->GetSpanContext().IsValid());
+    if (span->GetName() == "Attempt.grpc.testing.EchoTestService/Echo") {
+      attempt_span = span.get();
+      EXPECT_THAT(span->GetAttributes(),
+                  UnorderedElementsAre(
+                      Pair("transparent-retry", VariantWith<bool>(false)),
+                      Pair("previous-rpc-attempts", VariantWith<uint64_t>(0))));
+      // Verify outbound message event
+      const auto outbound_message_event =
+          std::find_if(span->GetEvents().begin(), span->GetEvents().end(),
+                       [](const SpanDataEvent& event) {
+                         return event.GetName() == "Outbound message";
+                       });
+      EXPECT_NE(outbound_message_event, span->GetEvents().end());
```

Comment:
Doesn't this need to be `ASSERT_NE()`?  I assume that if this expectation fails, the expectation on the next line will crash.

Same throughout.